### PR TITLE
WIP: Integrate FIP0013

### PR DESCRIPTION
**Pull request by arajasek**
Overhauls #5769 to be on v5 actors and go into the staging branch.

Currently this is pulling in https://github.com/filecoin-project/filecoin-ffi/pull/169, https://github.com/filecoin-project/go-paramfetch/pull/13 and https://github.com/filecoin-project/specs-actors/pull/1381, and implementing some plumbing logic for proof aggregation

TODO:
- [ ] Write some unit tests for CommitBatcher
- [ ] Write an integration test for commit batching
  - [ ] Test edge-cases
    - [ ] One porep fails
    - [ ] One commit expires? (deal start?)
    - [ ] Batch fails for whatever reason (e.g. out of gas)
- [ ] Use correct specs-actors commit (currently using `asr/temp`).

### CODE REVIEW

**Comment by BigLep** [comment]
2021-05-13 conversation:
1. Around FFIs: https://github.com/filecoin-project/go-state-types/commit/4aaed5f37509899cb438a761b501808ac0212529#diff-1db57c219b5669cc19920de5e805ebd15be6809e27771250bced48416867d80dR103
2. Where proofs are coming from:
 - Are regular PoReps 
 - We don't think we need a new API.  We "think" the logic is in the miner state machine.

**Comment by BigLep** [comment]
2021-05-17: for getting this to merge:
1. configuration
2. unit tests

File: node/config/def.go
Comment (vyzo) [inline node/config/def.go:247]: this cannot exceed 204 per FIP.

File: extern/sector-storage/ffiwrapper/types.go
Comment (Stebalien) [inline extern/sector-storage/ffiwrapper/types.go:44]: Really, neither this nor `GenerateWinningPoStSectorChallenge` belong here. Why do this?

File: extern/storage-sealing/commit_batch.go
Comment (GFZRZK) [inline extern/storage-sealing/commit_batch.go:187]: here can take at most `MaxAggregatedSectors` and proof size can't longer than `MaxAggregateProofSize`, if you don't defend here the `processBatch` may never work out a right result.

File: extern/sector-storage/ffiwrapper/types.go
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:44]: Yeah, we probably want a separate interface for these. This is here because the Sealer instance is wrapped into a Worker+SealingMgr, and it really makes no sense to put those 2 calls behind those layers of abstraction.

The `Verifier` interface is the other ready-to-use interface that's already plumbed into all the places we need those calls in.

File: extern/storage-sealing/commit_batch.go
Comment (DC-Byteark) [inline extern/storage-sealing/commit_batch.go:197]: This should be proofs, not infos.

File: api/test/window_post.go
Comment (magik6k) [inline api/test/window_post.go:160]: todo more asserts here

File: api/test/window_post.go
Comment (magik6k) [inline api/test/window_post.go:120]: Should be in a different file

File: extern/storage-sealing/states_sealing.go
Comment (magik6k) [inline extern/storage-sealing/states_sealing.go:228]: This can send events, but we ignore that in callers

File: storage/wdpost_run_test.go
Comment (magik6k) [inline storage/wdpost_run_test.go:150]: Not on the interface anymore

File: documentation/en/api-methods-miner.md
Comment (ZenGround0) [inline documentation/en/api-methods-miner.md:1563]: nit: add docs

File: documentation/en/cli-lotus-miner.md
Comment (ZenGround0) [inline documentation/en/cli-lotus-miner.md:1]: Most landed in #6145, only reviewing batching section which lgtm

File: extern/sector-storage/ffiwrapper/types.go
Comment (ZenGround0) [inline extern/sector-storage/ffiwrapper/types.go:47]: Reminder to do this

File: extern/sector-storage/mock/mock.go
Comment (ZenGround0) [inline extern/sector-storage/mock/mock.go:529]: nit: possible debug print

File: extern/sector-storage/mock/mock.go
Comment (ZenGround0) [inline extern/sector-storage/mock/mock.go:546]: nit: possible debug print

File: markets/storageadapter/ondealsectorcommitted.go
Comment (ZenGround0) [inline markets/storageadapter/ondealsectorcommitted.go:151]: suggestion: this feels like it belongs in a function `onSectorNumber(msg.Methods, msg.Params, res.DealID, cb)`

File: markets/storageadapter/ondealsectorcommitted.go
Comment (ZenGround0) [inline markets/storageadapter/ondealsectorcommitted.go:236]: again consider putting this in its own function to keep `OnDealSectorCommitted` more compact

File: node/config/def.go
Comment (ZenGround0) [inline node/config/def.go:91]: Are there docs describing these things?  I have a general idea what each field does from the names but it's not exactly obvious.

**Review by ZenGround0 — COMMENTED** [review]
About half way

File: extern/storage-sealing/commit_batch.go
Comment (vyzo) [inline extern/storage-sealing/commit_batch.go:102]: I think we should only start the timer after the first commit enters the queue.

File: extern/sector-storage/ffiwrapper/types.go
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:47]: Should be done outside of this PR after it lands

Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:47]: (warrants an issue)

File: chain/stmgr/forks.go
Comment (magik6k) [reply chain/stmgr/forks.go:1250]: Oh, I just assumed that /something/ is expecting this to be bumped based on the nv12 diff, might have been wrong

File: extern/sector-storage/ffiwrapper/types.go
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:47]: https://github.com/filecoin-project/lotus/issues/6326

File: chain/vm/gas.go
Comment (magik6k) [reply chain/vm/gas.go:163]: Unless @Kubuxu disagrees I'd say no

File: chain/stmgr/forks.go
Comment (magik6k) [reply chain/stmgr/forks.go:1250]: I guess the only good reason for it is version check in `chain/stmgr/forks.go:upgradeActorsV5Common`, I'm ok with dropping that

**Review by ZenGround0 — APPROVED** [review]
Integration LGTM

I have not yet reviewed the extensive storage-sealing changes.  If we want a thorough review from me on this it is going to take significant time to get up to speed because the storage fsm is totally new code to me.  That might be expected but I want to check in about it before moving forward.

File: extern/storage-sealing/commit_batch.go
Comment (arajasek) [inline extern/storage-sealing/commit_batch.go:37]: This can now be removed

**Review by arajasek — COMMENTED** [review]
Reviewed the storage logic, and it looks good. Ran a devnet and it seems to do the thing.

(Verbal green check mark from me, since it's """my PR""")

File: chain/stmgr/forks.go
Comment (ZenGround0) [reply chain/stmgr/forks.go:1250]: I defer to you and @arajasek

File: extern/storage-sealing/precommit_batch.go
Comment (GFZRZK) [inline extern/storage-sealing/precommit_batch.go:106]: Simply handle errors here will cause a deadlock. Amusing that we meet error in `processBatch` i.e.: an API error or meet no enough money. And it lasts till sector expired, then here will always return error from `sendMsg` since specs-actors will not allow a expired sector pass. I think when meet error, here should check every sector's validation and close and send the error to those sectors which will never success

Comment (GFZRZK) [reply extern/storage-sealing/precommit_batch.go:106]: @magik6k what's your opinion?

**Pull request merged**
